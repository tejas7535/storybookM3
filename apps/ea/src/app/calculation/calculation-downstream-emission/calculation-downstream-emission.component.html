@if (downstreamEmission.length > 1) {
  <!-- Mobile resolution -->
  <div class="sm:hidden">
    <div class="w-full overflow-hidden relative flex flex-col">
      @for (
        value of downstreamEmission;
        track value;
        let i = $index;
        let isFirst = $first;
        let isLast = $last
      ) {
        <div [ngClass]="{ hidden: i !== selectedIndex }">
          <div class="flex justify-center mb-2">
            @if (!isFirst) {
              <button
                mat-icon-button
                class="flex items-center !text-high-emphasis absolute left-0"
                (click)="previousItem()"
              >
                <mat-icon class="mr-1 text-outline-variant"
                  >arrow_back</mat-icon
                >
              </button>
            }
            <button
              mat-button
              class="flex items-center !text-high-emphasis bg-surface-container"
            >
              @if (!isFirst) {
                <mat-icon
                  class="mr-1"
                  [ngStyle]="{
                    color: chartColors[i % chartColors.length]
                  }"
                  >circle</mat-icon
                >
              }
              {{ value.id }}
            </button>

            @if (!isLast) {
              <button
                mat-icon-button
                class="flex items-center !text-high-emphasis absolute right-0"
                (click)="nextItem()"
              >
                <mat-icon class="mr-1 text-outline-variant"
                  >arrow_forward</mat-icon
                >
              </button>
            }
          </div>

          @if (i === selectedIndex) {
            <div
              contentSwipe
              (swipeLeft)="previousItem()"
              (swipeRight)="nextItem()"
              [@swipe]="getSwipeParams()"
            >
              <ea-report-emissions-values
                [co2Emission]="value.emission"
                [co2EmissionPercentage]="value.emissionPercentage"
                [operatingHours]="value.operatingTimeInHours"
                [downstreamError]="downstreamError"
              ></ea-report-emissions-values>
            </div>
          }
        </div>
      }
    </div>

    <div class="dots-container flex justify-center mt-2">
      @for (value of downstreamEmission; track value; let i = $index) {
        <span
          class="dot h-1.5 w-1.5 my-0 mx-1 bg-surface-container-highest inline-block rounded-full transition duration-500 easy"
          [ngClass]="{
            'bg-outline-variant h-2 w-2': i === selectedIndex,
            'bg-surface-container-highest': i !== selectedIndex
          }"
        ></span>
      }
    </div>
  </div>

  <!-- Desktop resolution -->

  <div class="mb-3 hidden sm:flex flex-wrap">
    @for (
      loadcase of downstreamEmission;
      track loadcase.id;
      let i = $index;
      let isFirst = $first
    ) {
      <button
        (click)="selectItem(i)"
        mat-button
        class="flex items-center !text-high-emphasis"
        [ngClass]="{ 'bg-surface-container': i === selectedIndex }"
      >
        @if (!isFirst) {
          <mat-icon
            class="mr-1"
            [ngStyle]="{
              color: chartColors[i % chartColors.length]
            }"
            >circle</mat-icon
          >
        }
        {{ loadcase.id }}
      </button>
    }
  </div>
}

@if (downstreamEmission.length > 0) {
  <ea-report-emissions-values
    class="sm:block"
    [ngClass]="{
      hidden: downstreamEmission.length > 1
    }"
    [co2Emission]="downstreamEmission[selectedIndex].emission"
    [co2EmissionPercentage]="
      downstreamEmission[selectedIndex].emissionPercentage
    "
    [operatingHours]="downstreamEmission[selectedIndex].operatingTimeInHours"
    [downstreamError]="downstreamError"
  ></ea-report-emissions-values>
}
