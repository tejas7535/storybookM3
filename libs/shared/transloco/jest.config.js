module.exports = {
  name: 'shared-transloco',
  preset: '../../../jest.config.js',
  coverageDirectory: '../../../coverage/libs/shared/transloco',
  snapshotSerializers: [
    'jest-preset-angular/build/AngularNoNgAttributesSnapshotSerializer.js',
    'jest-preset-angular/build/AngularSnapshotSerializer.js',
    'jest-preset-angular/build/HTMLCommentSerializer.js'
  ]
};
