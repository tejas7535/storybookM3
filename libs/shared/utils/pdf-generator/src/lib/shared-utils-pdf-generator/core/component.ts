/* eslint-disable max-lines */
import { jsPDF } from 'jspdf';

import { FontOptions } from './format';
import { PDFDocument } from './pdfdocument';
import { Rect } from './rect';

const DEFAULT_EMPTY_RECT = new Rect(0, 0, 0, 0);

export interface PageContext {
  position: 'top' | 'bottom';
  pageNumber: number;
  totalPages: number;
}

interface TextOptions {
  fontOptions?: FontOptions;
  link?: string;
  textOptions?: Parameters<jsPDF['text']>[3];
}

export abstract class Component {
  protected _doc?: jsPDF;

  protected bounds: Rect = DEFAULT_EMPTY_RECT;

  protected pageContext?: PageContext;

  protected _pdfDoc?: PDFDocument;
  private renderDebug = false;

  public setDocument(document: PDFDocument | undefined): void {
    if (document?.pdfDoc) {
      this._pdfDoc = document;
      this._doc = document.pdfDoc;
    }
  }

  public setBounds(bounds: Rect) {
    this.bounds = bounds;
  }

  public setPageContext(context: PageContext): void {
    this.pageContext = context;
  }

  /**
   * When set to true, it draws and outline around the bounds region
   * to allow for a visual reference
   **/
  public setDebug(debug: boolean) {
    this.renderDebug = debug;
  }

  /**
   * Evaluate is called before the actual rendering takes place.
   * It is used to understand the size of the component for layouting
   * purposes before actually drawing them.
   *
   * The main concern of this function is laying out along a horizontal axis.
   * If a component does not fit into the bounding rectangle,
   *
   *
   * Wrapping may be optional, the wrapping behavior i
   * @returns
   *   - boolean indicating the fit
   *   - number indicating the actually required vertical space
   *   - partial component that fits into the bounding rect (undefined if all of it fits)
   *   - partial component that does not fit into the bounding rect (undefined if all of it fits)
   **/
  public evaluate(bounds: Rect): [boolean, number, Component?, Component?] {
    this._pdfDoc?.reset();
    this.bounds = bounds;

    return [false, 0];
  }

  public render(): void {
    if (this.bounds === DEFAULT_EMPTY_RECT) {
      throw new Error('Called render before setting rendering bounds.');
    }
    this._pdfDoc?.reset();
    this.drawDebug();
  }

  /**
   * This is a wrapper function around the jsPDF#getTextDimensions.
   * Unfortunately the the documented types provided by the jsPDF library seem to be wrong
   * as supplying the font argument as a string results in runtime errors, suggesting the font
   * was not resvoled properly.
   *
   * Instead, this helper function handles the temporary application of font styles outside the jsPDF
   * library.
   *
   * Also, providing the different parameters (custom font, custom font size) seems to work
   * does not produce a reliable behavior when dealing with multiline strings on the document
   *
   **/
  protected getTextDimensions(
    text: string,
    options?: FontOptions & Parameters<jsPDF['getTextDimensions']>[1]
  ) {
    if (!this._doc) {
      throw new Error('Called before setting pdf document to component');
    }

    let resetFn;
    if (options?.fontFamily || options?.fontSize || options?.font) {
      const tempStyles: FontOptions = {
        fontFamily:
          options.font ||
          options.fontFamily ||
          options.fontFamily ||
          this._doc.getFont().fontName,
        fontSize: options.fontSize || this._doc.getFontSize(),
      };
      resetFn = this.temporaryFontStyle(tempStyles);
    }

    const dimensions = this._doc.getTextDimensions(
      text,
      options?.maxWidth ? { maxWidth: options.maxWidth } : undefined
    );

    if (resetFn) {
      resetFn();
    }

    return dimensions;
  }

  protected getLineHeight(fontFormat?: FontOptions) {
    const _doc = this.assertDoc();
    let resetFontStyle;
    if (fontFormat) {
      resetFontStyle = this.temporaryFontStyle(fontFormat);
    }
    const height = _doc.getFontSize() / _doc.getLineHeightFactor();

    if (resetFontStyle) {
      resetFontStyle();
    }

    return height;
  }

  protected getStringWidth(content: string) {
    const doc = this.assertDoc();

    return doc.getTextWidth(content);
  }

  protected temporaryFontStyle(fontFormat: FontOptions) {
    if (!this._doc) {
      throw new Error('Called before setting pdf document to component');
    }
    const previousFont = this._doc.getFont();
    const previousFontSize = this._doc.getFontSize();

    const resetClosure = () => {
      if (!this._doc) {
        throw new Error(
          'Reset closure has been called, but jsPDF document is undefined at component scope'
        );
      }
      this._doc.setFont(
        previousFont.fontName,
        previousFont.fontStyle || 'normal'
      );
      this._doc?.setFontSize(previousFontSize);
    };

    if (fontFormat.fontSize) {
      this._doc.setFontSize(fontFormat.fontSize);
    }

    if (fontFormat.fontFamily) {
      this._doc.setFont(
        fontFormat.fontFamily,
        fontFormat.fontStyle || 'normal'
      );
    }

    return resetClosure;
  }

  /**
   * Draw a piece of text to the PDF page while dealing with some of the oddities with the coorodinates
   *
   * @param x coordinate of the top left corner
   * @param y coordinate of the top left corner
   * @param options options to specifiy the formatting of the text
   **/
  protected text(
    x: number,
    y: number,
    text: string | string[],
    options?: TextOptions
  ) {
    if (!this._doc) {
      throw new Error('component.text() missing reference document');
    }
    let resetStyles;

    if (options?.fontOptions) {
      resetStyles = this.temporaryFontStyle(options.fontOptions);
    }
    let textHeight = 0;
    if (Array.isArray(text)) {
      // TODO: find way to reliably get the height of a multiline text
      return;
    } else {
      textHeight = this._doc.getTextDimensions(text).h;
    }
    if (options?.link) {
      this._doc.textWithLink(text, x, y + textHeight, { url: options?.link });
    } else {
      this._doc.text(text, x, y + textHeight, options?.textOptions);
    }

    if (resetStyles) {
      resetStyles();
    }
  }

  /*
   * Helper method to set the text color of the document
   * @param color the color to set
   * @returns a function that resets the text color to the original value
   **/
  protected setTextColor(color: string): () => void {
    const doc = this.assertDoc();
    const originalTextColor = doc.getTextColor();
    doc.setTextColor(color);

    const resetClosure = () => {
      doc.setTextColor(originalTextColor);
    };

    return resetClosure;
  }

  /**
   * Helper method to draw an image to the page.
   * The scaling behaviour differs based on the supplied arguments
   *
   * NOTE: if both the width and the height parameters are undefined, the image will be drawn in its original dimensions
   *
   * @param imageData - raw image data as defined by jsPDF
   * @param x starting coordinate
   * @param y starting coordinate
   * @param width the width of the image. If this argument is undefined, the image is scaled to the expected height while retaining the original images aspect ratio
   * @param height the height of the image. If this argument is undefined, the image is caled to the expected with while retaining the original image aspect ratio
   *
   * @throws This function can only be called in the generate scope. If it is called before the jsPDF document context is set, the the function will throw an error
   **/
  protected image(
    imageData: Parameters<jsPDF['getImageProperties']>[0],
    x: number,
    y: number,
    width?: number,
    height?: number
  ) {
    const pdf = this.assertDoc();
    const [imageWidth, imageHeight] = this.scaleImage(imageData, width, height);
    pdf.addImage(imageData, x, y, imageWidth!, imageHeight!);
  }

  /**
   * Utilility function to get aspect ratio preserving scaling done for images supplied to the function. Scaling can be done either by supplying the height
   * or the width of the image.
   *
   * If both values passed are undefined, the native dimensions of the provided image resource are returned
   *
   * @param width (optional) the target width of the image. Will be derived from the supplied hight if undefined, retaining the aspect ratio
   * @param height (optional) the target height of the image. Will be derived from the supplied widht if undefined, retaining the aspect ratio
   * @returns an array with the resulting image with at index 0 and the height at 0
   **/
  protected scaleImage(
    imageData: Parameters<jsPDF['getImageProperties']>[0],
    width?: number,
    height?: number
  ): [number, number] {
    const pdf = this.assertDoc();

    const imageProperties = pdf.getImageProperties(imageData);
    let imageWidth: number;
    let imageHeight: number;

    if (width && height) {
      imageWidth = width;
      imageHeight = height;
    } else if (!width && !height) {
      imageWidth = imageProperties.width;
      imageHeight = imageProperties.height;
    } else if (width && !height) {
      imageWidth = width;
      imageHeight = imageProperties.height / (imageProperties.width / width);
    } else if (!width && height) {
      imageHeight = height;
      imageWidth = imageProperties.width / (imageProperties.height / height);
    }

    return [imageWidth!, imageHeight!];
  }

  protected hline(y: number, color?: string) {
    const pdf = this.assertDoc();
    let resetColor: string | undefined;
    if (color) {
      resetColor = pdf.getDrawColor();
      pdf.setDrawColor(color);
    }
    pdf.line(0, y, pdf.internal.pageSize.getWidth(), y);
    if (resetColor) {
      pdf.setDrawColor(resetColor);
    }
  }

  protected assertPageContext(): PageContext {
    if (!this.pageContext) {
      throw new Error(
        'Page context is not avaiable. The page context property is only set on repeat elements like headers or footers'
      );
    }

    return this.pageContext;
  }

  protected assertDoc(): jsPDF {
    if (!this._doc) {
      throw new Error(
        'jsPDF object missing from component scope. This function was called outside the #generate scope of the document'
      );
    }

    return this._doc;
  }
  /**
   * Using the jsPDF#getTextDimensions() currently only works reliably for single line strings.
   * In instances where strings span over mutliple lines of text, the value caulculated is unreliable.
   *
   * This method of calculating is basedo n a pull request to the jsPDF library that is currently waiting
   * for review
   * @see https://github.com/parallax/jsPDF/pull/3814
   *
   * @param text the text to calculate the height of
   * @param maxWidth the maximum with the text is allowed to take. Splitting is done using the doc.splitTextToSize option
   * @param fontFormat the optional font format that should be applied. If undefined, the current document format is used
   **/
  protected getMultilineTextHeight(
    text: string,
    maxWidth: number,
    fontFormat?: FontOptions
  ) {
    const doc = this.assertDoc();
    let resetFn;

    if (fontFormat) {
      resetFn = this.temporaryFontStyle(fontFormat);
    }

    const splitUpText = doc.splitTextToSize(text, maxWidth) as string[];
    const lineHeightFacotor = doc.getLineHeightFactor();
    const fontSize = doc.getFontSize();

    const multilineTextHeight =
      (splitUpText.length * fontSize * lineHeightFacotor -
        fontSize * (lineHeightFacotor - 1)) /
      doc.internal.scaleFactor;

    if (resetFn) {
      resetFn();
    }

    return multilineTextHeight;
  }

  /**
   * Placing the text always requires the top left coordinate, making alignment
   * of the text to the right hard.
   **/
  protected alignRight(x: number, text: string, font?: FontOptions): number {
    return x - this.getTextDimensions(text, font).w;
  }

  protected vcenter(
    y: number,
    height: number,
    text: string,
    maxWidth: number,
    fontOptions?: FontOptions
  ) {
    const textHeight = this.getMultilineTextHeight(text, maxWidth, fontOptions);

    return y + height / 2 - textHeight / 2;
  }

  /**
   * Utility function to help with centering a text horizontally in a given box
   *
   * @param width of the box to center in
   * @param text to center inside
   * @param fontOptions (optional) to apply when calculating the width
   *
   * @returns the x coordinate of the top left text anchor
   **/
  protected hcenter(
    x: number,
    width: number,
    text: string,
    fontOptions?: FontOptions
  ) {
    const textWidth = this.getTextDimensions(text, fontOptions).w;

    return x + (width / 2 - textWidth / 2);
  }

  private drawDebug() {
    if (this.bounds === DEFAULT_EMPTY_RECT) {
      throw new Error('Called render before setting rendering bounds.');
    }
    if (this.renderDebug) {
      this._doc?.rect(
        this.bounds.x,
        this.bounds.y,
        this.bounds.width,
        this.bounds.height
      );
    }
  }
}
