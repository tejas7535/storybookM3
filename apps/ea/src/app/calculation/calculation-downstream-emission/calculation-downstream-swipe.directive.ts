import { Directive, ElementRef, OnDestroy, output } from '@angular/core';

@Directive({
  // eslint-disable-next-line @angular-eslint/directive-selector
  selector: '[contentSwipe]',
  standalone: true,
})
export class CalculationDowstreamSwipeDirective implements OnDestroy {
  swipeLeft = output();
  swipeRight = output();

  private touchStartX: number;

  constructor(private readonly el: ElementRef) {
    this.el.nativeElement.addEventListener(
      'touchstart',
      this.onTouchStart.bind(this),
      { passive: true }
    );

    this.el.nativeElement.addEventListener(
      'touchend',
      this.onTouchEnd.bind(this),
      { passive: true }
    );
  }

  ngOnDestroy(): void {
    this.el.nativeElement.removeEventListener('touchstart', this.onTouchStart);
    this.el.nativeElement.removeEventListener('touchend', this.onTouchEnd);
  }

  onTouchStart(event: TouchEvent): void {
    this.touchStartX = event.touches[0].clientX;
  }

  onTouchEnd(event: TouchEvent): void {
    const deltaX = event.changedTouches[0].clientX - this.touchStartX;
    const swipeThreshold = 30; // Minimum swipe distance in pixels

    if (Math.abs(deltaX) > swipeThreshold) {
      if (deltaX < 0) {
        this.swipeRight.emit();
      } else {
        this.swipeLeft.emit();
      }
    }
  }
}
