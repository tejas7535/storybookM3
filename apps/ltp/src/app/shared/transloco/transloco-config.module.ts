import { APP_INITIALIZER, ModuleWithProviders, NgModule } from '@angular/core';

import {
  getBrowserLang,
  HashMap,
  TRANSLOCO_CONFIG,
  TranslocoConfig,
  TranslocoModule,
  TranslocoService
} from '@ngneat/transloco';

import { environment } from '../../../environments/environment';
import { translocoLoader } from './transloco.loader';

const translocoConfig: TranslocoConfig = {
  availableLangs: ['en', 'de'],
  defaultLang: 'en',
  fallbackLang: 'en',
  prodMode: environment.production,
  reRenderOnLangChange: true,
  flatten: {
    aot: environment.production
  }
};

const preloadLanguage = (transloco: TranslocoService): Function => {
  return async (): Promise<HashMap> => {
    const lang = getBrowserLang();

    transloco.setActiveLang(lang);

    return transloco.load(lang).toPromise();
  };
};

const preLoad = {
  provide: APP_INITIALIZER,
  multi: true,
  useFactory: preloadLanguage,
  deps: [TranslocoService]
};

@NgModule({
  imports: [TranslocoModule],
  exports: [TranslocoModule]
})
export class TranslocoConfigModule {
  static forRoot(): ModuleWithProviders {
    return {
      ngModule: TranslocoConfigModule,
      providers: [
        {
          provide: TRANSLOCO_CONFIG,
          useValue: translocoConfig
        },
        translocoLoader,
        preLoad
      ]
    };
  }
}
