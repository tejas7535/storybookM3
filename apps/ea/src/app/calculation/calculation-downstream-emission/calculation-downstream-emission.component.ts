import { animate, style, transition, trigger } from '@angular/animations';
import { CommonModule } from '@angular/common';
import {
  ChangeDetectionStrategy,
  Component,
  ElementRef,
  EventEmitter,
  Input,
  OnDestroy,
  OnInit,
  Output,
} from '@angular/core';
import { MatButtonModule } from '@angular/material/button';
import { MatIconModule } from '@angular/material/icon';

import { CO2EmissionResult } from '@ea/core/store/selectors/calculation-result/calculation-result-report.selector';
import { CHARTS_COLORS } from '@ea/shared/constants/charts-colors';
import { TranslocoService } from '@jsverse/transloco';

import { SharedTranslocoModule } from '@schaeffler/transloco';

import { ReportCo2EmissionsValuesComponent } from '../report-co2-emissions-values/report-co2-emissions-values.component';

interface ResultItem {
  id: string;
  emission: number;
  emissionPercentage: number;
  operatingTimeInHours: number;
}

@Component({
  selector: 'ea-calculation-downstream-emission',
  templateUrl: './calculation-downstream-emission.component.html',
  standalone: true,
  changeDetection: ChangeDetectionStrategy.OnPush,
  imports: [
    CommonModule,
    ReportCo2EmissionsValuesComponent,
    SharedTranslocoModule,
    MatIconModule,
    MatButtonModule,
  ],
  animations: [
    trigger('swipe', [
      transition(
        ':enter',
        [
          style({ transform: '{{enterTransform}}' }),
          animate('300ms ease-out', style({ transform: 'translateX(0)' })),
        ],
        { params: { enterTransform: 'translateX(100%)' } }
      ),
      transition(
        ':leave',
        [animate('300ms ease-out', style({ transform: '{{leaveTransform}}' }))],
        { params: { leaveTransform: 'translateX(-100%)' } }
      ),
    ]),
  ],
})
export class CalculationDownstreamEmissionComponent
  implements OnInit, OnDestroy
{
  @Input()
  public downstreamError?: string;

  @Output() selectedIndexChange: EventEmitter<number> = new EventEmitter();

  public _downstreamEmission: ResultItem[] = [];
  public readonly chartColors = CHARTS_COLORS;
  public swipeDirection: 'left' | 'right' = 'left';

  private _selectedIndex: number | undefined;

  private touchStartX = 0;
  private touchEndX = 0;

  constructor(
    private readonly translocoService: TranslocoService,
    private readonly el: ElementRef
  ) {}

  get selectedIndex(): number {
    return this._selectedIndex;
  }

  @Input() set selectedIndex(index: number | undefined) {
    this._selectedIndex = index ?? 0;
  }

  @Input() set downstreamEmission(
    emission: CO2EmissionResult['co2_downstream']
  ) {
    const operatingTimeInHours = emission?.loadcases?.reduce(
      (total, loadcase) => total + loadcase.operatingTimeInHours,
      0
    );
    const emissionItems: ResultItem[] = [
      {
        id: this.translocoService.translate(
          'calculationResultReport.co2Emissions.totalDownstreamButton'
        ),
        emission: emission?.emission,
        emissionPercentage: emission?.emissionPercentage,
        operatingTimeInHours,
      },
    ];

    if (emission.loadcases?.length > 1) {
      emissionItems.push(...emission.loadcases);
    }

    this._downstreamEmission = emissionItems;
  }

  // eslint-disable-next-line @typescript-eslint/member-ordering
  get downstreamEmission(): ResultItem[] {
    return this._downstreamEmission;
  }

  ngOnInit() {
    this.el.nativeElement.addEventListener(
      'touchstart',
      this.onTouchStart.bind(this),
      { passive: true }
    );
    this.el.nativeElement.addEventListener(
      'touchmove',
      this.onTouchMove.bind(this),
      { passive: true }
    );
    this.el.nativeElement.addEventListener(
      'touchend',
      this.onTouchEnd.bind(this),
      { passive: true }
    );
  }

  onTouchStart(event: TouchEvent): void {
    this.touchStartX = event.changedTouches[0].screenX;
  }

  onTouchMove(event: TouchEvent): void {
    this.touchEndX = event.changedTouches[0].screenX;
  }

  onTouchEnd(): void {
    const swipeThreshold = 30; // Minimum swipe distance in pixels

    if (Math.abs(this.touchEndX - this.touchStartX) > swipeThreshold) {
      if (this.touchEndX < this.touchStartX) {
        this.nextItem();
      } else if (this.touchEndX > this.touchStartX) {
        this.previousItem();
      }
    }
  }

  previousItem(): void {
    const previous = this.selectedIndex - 1;
    if (previous >= 0) {
      this.selectedIndexChange.emit(previous);
      this.swipeDirection = 'right';
    } else {
      this.selectedIndexChange.emit(0);
    }
  }

  nextItem(): void {
    const next = this.selectedIndex + 1;

    if (next < this._downstreamEmission.length) {
      this.swipeDirection = 'left';
      this.selectedIndexChange.emit(next);
    }
  }

  selectItem(index: number) {
    this.selectedIndexChange.emit(index);
  }

  getSwipeParams() {
    const directions = {
      left: {
        enterTransform: 'translateX(100%)',
        leaveTransform: 'translateX(-100%)',
      },
      right: {
        enterTransform: 'translateX(-100%)',
        leaveTransform: 'translateX(100%)',
      },
    };

    return {
      value: '',
      params: directions[this.swipeDirection],
    };
  }

  ngOnDestroy() {
    this.el.nativeElement.removeEventListener('touchstart', this.onTouchStart);
    this.el.nativeElement.removeEventListener('touchmove', this.onTouchMove);
    this.el.nativeElement.removeEventListener('touchend', this.onTouchEnd);
  }
}
