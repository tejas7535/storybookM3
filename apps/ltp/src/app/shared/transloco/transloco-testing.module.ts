import { TranslocoConfig, TranslocoTestingModule } from '@ngneat/transloco';

import * as en from '../../../assets/i18n/en.json';

export const getTranslocoModule = (config: Partial<TranslocoConfig> = {}) =>
  TranslocoTestingModule.withLangs(
    { en },
    {
      availableLangs: ['en'],
      defaultLang: 'en',
      ...config
    }
  );
