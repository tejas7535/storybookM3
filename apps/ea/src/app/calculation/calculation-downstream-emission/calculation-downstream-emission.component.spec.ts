import { createComponentFactory, Spectator } from '@ngneat/spectator/jest';

import { provideTranslocoTestingModule } from '@schaeffler/transloco/testing';

import { CalculationDownstreamEmissionComponent } from './calculation-downstream-emission.component';

describe('CalculationDownstreamEmissionComponent', () => {
  let spectator: Spectator<CalculationDownstreamEmissionComponent>;
  let component: CalculationDownstreamEmissionComponent;
  const createComponent = createComponentFactory({
    component: CalculationDownstreamEmissionComponent,
    imports: [provideTranslocoTestingModule({ en: {} })],
  });

  beforeEach(() => {
    spectator = createComponent();
    component = spectator.component;
  });

  it('should create the component', () => {
    expect(component).toBeTruthy();
  });

  describe('when setting up emission', () => {
    beforeEach(() => {
      const co2_downstream = {
        emission: 123,
        emissionPercentage: 50,
        loadcases: [
          {
            id: 'Loadcase 1',
            emission: 0.49,
            unit: 'kg',
            emissionPercentage: 7.89,
            operatingTimeInHours: 150,
          },
          {
            id: 'Loadcase 2',
            emission: 0.51,
            unit: 'kg',
            emissionPercentage: 8.01,
            operatingTimeInHours: 250,
          },
        ],
      };
      spectator.setInput('selectedIndex', undefined);
      spectator.setInput('downstreamEmission', co2_downstream);
    });

    it('should set downstream emission', () => {
      expect(component.downstreamEmission).toMatchSnapshot();
    });

    it('should emit selectedIndexChange with the correct index', () => {
      const testIndex = 1;
      jest.spyOn(component.selectedIndexChange, 'emit');

      component.selectItem(testIndex);

      expect(component.selectedIndexChange.emit).toHaveBeenCalledWith(
        testIndex
      );
    });

    describe('when swiping loadcases list', () => {
      it('should handle touch start', () => {
        const touchEvent = new TouchEvent('touchstart', {
          changedTouches: [
            {
              identifier: 0,
              target: window,
              screenX: 100,
              screenY: 0,
            } as unknown as Touch,
          ],
        });

        component.onTouchStart(touchEvent);
        expect(component['touchStartX']).toBe(100);
      });

      it('should handle touch move', () => {
        const touchEvent = new TouchEvent('touchmove', {
          changedTouches: [
            {
              identifier: 0,
              target: window,
              screenX: 150,
              screenY: 0,
            } as unknown as Touch,
          ],
        });

        component.onTouchMove(touchEvent);
        expect(component['touchEndX']).toBe(150);
      });

      it('should handle touch end and swipe left', () => {
        component['touchStartX'] = 100;
        component['touchEndX'] = 50;

        jest.spyOn(component, 'nextItem');
        jest.spyOn(component, 'previousItem');

        component.onTouchEnd();

        expect(component.nextItem).toHaveBeenCalled();
        expect(component.previousItem).not.toHaveBeenCalled();
      });

      it('should handle touch end and swipe right', () => {
        component['touchStartX'] = 50;
        component['touchEndX'] = 100;

        jest.spyOn(component, 'nextItem');
        jest.spyOn(component, 'previousItem');

        component.onTouchEnd();

        expect(component.previousItem).toHaveBeenCalled();
        expect(component.nextItem).not.toHaveBeenCalled();
      });

      it('should not trigger swipe if distance is less than threshold', () => {
        component['touchStartX'] = 100;
        component['touchEndX'] = 120;

        jest.spyOn(component, 'nextItem');
        jest.spyOn(component, 'previousItem');

        component.onTouchEnd();

        expect(component.nextItem).not.toHaveBeenCalled();
        expect(component.previousItem).not.toHaveBeenCalled();
      });

      it('should return correct swipe params for left direction', () => {
        component.swipeDirection = 'left';
        const result = component.getSwipeParams();
        expect(result).toEqual({
          value: '',
          params: {
            enterTransform: 'translateX(100%)',
            leaveTransform: 'translateX(-100%)',
          },
        });
      });

      it('should return correct swipe params for right direction', () => {
        component.swipeDirection = 'right';
        const result = component.getSwipeParams();
        expect(result).toEqual({
          value: '',
          params: {
            enterTransform: 'translateX(-100%)',
            leaveTransform: 'translateX(100%)',
          },
        });
      });
    });

    it('should emit previous index', () => {
      component.selectedIndex = 1;
      jest.spyOn(component.selectedIndexChange, 'emit');

      component.previousItem();

      expect(component.selectedIndexChange.emit).toHaveBeenCalledWith(0);
    });

    it('should emit zero if already at first index', () => {
      component.selectedIndex = 0;
      jest.spyOn(component.selectedIndexChange, 'emit');

      component.previousItem();

      expect(component.selectedIndexChange.emit).toHaveBeenCalledWith(0);
    });

    it('should emit next index', () => {
      component.selectedIndex = 1;
      jest.spyOn(component.selectedIndexChange, 'emit');

      component.nextItem();

      expect(component.selectedIndexChange.emit).toHaveBeenCalledWith(2);
    });

    it('should not emit if already at last index', () => {
      component.selectedIndex = 3;

      jest.spyOn(component.selectedIndexChange, 'emit');

      component.nextItem();

      expect(component.selectedIndexChange.emit).not.toHaveBeenCalled();
    });
  });

  describe('when there is not emission', () => {
    beforeEach(() => {
      const co2_downstream = {
        emission: 0,
        emissionPercentage: 0,
      };

      spectator.setInput('selectedIndex', undefined);
      spectator.setInput('downstreamEmission', co2_downstream);
    });

    it('should handle it gracefully', () => {
      expect(component.downstreamEmission).toMatchSnapshot();
    });
  });

  it('should add event listeners on ngOnInit', () => {
    const addEventListenerSpy = jest.spyOn(
      component['el'].nativeElement,
      'addEventListener'
    );

    component.ngOnInit();

    expect(addEventListenerSpy).toHaveBeenCalledWith(
      'touchstart',
      expect.any(Function),
      { passive: true }
    );
    expect(addEventListenerSpy).toHaveBeenCalledWith(
      'touchmove',
      expect.any(Function),
      { passive: true }
    );
    expect(addEventListenerSpy).toHaveBeenCalledWith(
      'touchend',
      expect.any(Function),
      { passive: true }
    );
  });

  it('should remove event listeners on ngOnDestroy', () => {
    const removeEventListenerSpy = jest.spyOn(
      component['el'].nativeElement,
      'removeEventListener'
    );

    component.ngOnDestroy();

    expect(removeEventListenerSpy).toHaveBeenCalledWith(
      'touchstart',
      expect.any(Function)
    );
    expect(removeEventListenerSpy).toHaveBeenCalledWith(
      'touchmove',
      expect.any(Function)
    );
    expect(removeEventListenerSpy).toHaveBeenCalledWith(
      'touchend',
      expect.any(Function)
    );
  });
});
