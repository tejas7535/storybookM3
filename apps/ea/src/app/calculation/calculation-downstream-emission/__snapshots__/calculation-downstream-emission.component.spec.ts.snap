// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CalculationDownstreamEmissionComponent when setting up emission should set downstream emission 1`] = `
Array [
  Object {
    "emission": 123,
    "emissionPercentage": 50,
    "id": "calculationResultReport.co2Emissions.totalDownstreamButton",
    "operatingTimeInHours": 400,
  },
  Object {
    "emission": 0.49,
    "emissionPercentage": 7.89,
    "id": "Loadcase 1",
    "operatingTimeInHours": 150,
    "unit": "kg",
  },
  Object {
    "emission": 0.51,
    "emissionPercentage": 8.01,
    "id": "Loadcase 2",
    "operatingTimeInHours": 250,
    "unit": "kg",
  },
]
`;

exports[`CalculationDownstreamEmissionComponent when there is not emission should handle it gracefully 1`] = `
Array [
  Object {
    "emission": 0,
    "emissionPercentage": 0,
    "id": "calculationResultReport.co2Emissions.totalDownstreamButton",
    "operatingTimeInHours": undefined,
  },
]
`;
